File: conway-ruby/lib/conway/game_model.rb
require_relative 'game_board'

module Conway
  class GameModel
    attr_reader :grid_state, :start_row, :start_col, :end_row, :end_col

    def initialize(max_grid_size)
      @max_grid_size = max_grid_size
      @game_board = Conway::GameBoard.new(max_grid_size)
      @grid_state = @game_board.grid
    end

    def populate_game_board(input_pattern)
      height = height_input_pattern(input_pattern)
      width = width_input_pattern(input_pattern)
      max_dimension_of_input_pattern = [height, width].max
      move_origin_to_point = 0
      if max_dimension_of_input_pattern + 1 < @max_grid_size
        move_origin_to_point = (@max_grid_size - max_dimension_of_input_pattern) / 2
      end
      @game_board.initialize_pattern(input_pattern, move_origin_to_point, move_origin_to_point)
      @start_row = move_origin_to_point
      @start_col = move_origin_to_point
      @end_row = @start_row + height
      @end_col = @start_col + width
    end

    def next_iteration
      temporary_grid = Array.new(@max_grid_size) { Array.new(@max_grid_size, ' ') }
      @flag_extend_left = false
      @flag_extend_right = false
      @flag_extend_top = false
      @flag_extend_bottom = false
      (0..@max_grid_size - 1).each do |row|
        (0..@max_grid_size - 1).each do |col|
          live_adjacent_cells = @game_board.number_of_live_adjacent_cells(row, col)

          temporary_grid[row][col] = '#' if @grid_state[row][col] == ' ' && live_adjacent_cells == 3

          if @grid_state[row][col] == '#' && (live_adjacent_cells >= 2 && live_adjacent_cells < 4)
            temporary_grid[row][col] = '#'
          end

          extend_grid(temporary_grid, row, col)
        end
      end
      change_start_end_coordinates
      @game_board.update_grid(temporary_grid)
    end

    def height_input_pattern(input_pattern)
      input_pattern.length
    end

    def width_input_pattern(input_pattern)
      width = 0
      (0..input_pattern.length - 1).each do |row|
        width = [width, input_pattern[row].length].max
      end
      width
    end

    def extend_grid(temporary_grid, row, col)
      @flag_extend_top = true if temporary_grid[row][col] == '#' && row == @start_row - 1
      @flag_extend_bottom = true if temporary_grid[row][col] == '#' && row == @end_row
      @flag_extend_left = true if temporary_grid[row][col] == '#' && col == @start_col - 1
      @flag_extend_right = true if temporary_grid[row][col] == '#' && col == @end_col
    end

    def change_start_end_coordinates
      @start_row -= 1 if @flag_extend_top
      @start_col -= 1 if @flag_extend_left
      @end_col += 1 if @flag_extend_right
      @end_row += 1 if @flag_extend_bottom
    end

    private :height_input_pattern, :width_input_pattern, :extend_grid, :change_start_end_coordinates
  end
end

File: conway-java/src/main/java/com/gojek/pasti/conway/GameModel.java
package com.gojek.pasti.conway;

import java.util.ArrayList;

public class GameModel {
  private final GameBoard gameBoard;
  private final int maxGridSize;
  public char[][] gridState;
  public int startRow;
  public int startCol;
  public int endRow;
  public int endCol;

  public GameModel(int maxGridSize) {
    this.maxGridSize = maxGridSize;
    gameBoard = new GameBoard(maxGridSize);
    gridState = gameBoard.getCurrentGrid();
  }

  public void populateGameBoard(ArrayList<ArrayList<Character>> inputPattern) {
    int height = heightInputPattern(inputPattern);
    int width = widthInputPattern(inputPattern);
    int maxDimensionOfInputPattern = Math.max(height, width);
    int moveOriginToPoint = 0;
    if (maxDimensionOfInputPattern + 1 < maxGridSize) {
      moveOriginToPoint = (maxGridSize - maxDimensionOfInputPattern) / 2;
    }
    gameBoard.initializePattern(inputPattern, moveOriginToPoint, moveOriginToPoint);
    startRow = moveOriginToPoint;
    startCol = moveOriginToPoint;
    endRow = startRow + height;
    endCol = startCol + width;
  }

  public void nextIteration() {
    boolean flagExtendLeft = false;
    boolean flagExtendRight = false;
    boolean flagExtendTop = false;
    boolean flagExtendBottom = false;
    char[][] temporaryBoard = new char[maxGridSize][maxGridSize];

    for (int row = 0; row < maxGridSize; row++) {
      for (int col = 0; col < maxGridSize; col++) {
        if (gridState[row][col] == ' ') {
          if (gameBoard.numberOfLiveAdjacentCells(row, col) == 3) {
            temporaryBoard[row][col] = '#';
          } else {
            temporaryBoard[row][col] = ' ';
          }
        } else {
          if (gameBoard.numberOfLiveAdjacentCells(row, col) == 2
              || gameBoard.numberOfLiveAdjacentCells(row, col) == 3) {
            temporaryBoard[row][col] = '#';
          } else {
            temporaryBoard[row][col] = ' ';
          }
        }

        if (temporaryBoard[row][col] == '#' && row == startRow - 1) flagExtendTop = true;
        if (temporaryBoard[row][col] == '#' && row == endRow) flagExtendBottom = true;
        if (temporaryBoard[row][col] == '#' && col == startCol - 1) flagExtendLeft = true;
        if (temporaryBoard[row][col] == '#' && col == endCol) flagExtendRight = true;
      }
    }
    extendGrid(flagExtendLeft, flagExtendRight, flagExtendTop, flagExtendBottom);
    gameBoard.updateGrid(temporaryBoard);
  }

  private int heightInputPattern(ArrayList<ArrayList<Character>> inputPattern) {
    return inputPattern.size();
  }

  private int widthInputPattern(ArrayList<ArrayList<Character>> inputPattern) {
    int width = 0;
    for (ArrayList<Character> inputPatternRow : inputPattern) {
      width = Math.max(width, inputPatternRow.size());
    }
    return width;
  }

  private void extendGrid(
      boolean flagLeft, boolean flagRight, boolean flagTop, boolean flagBottom) {
    if (flagTop) startRow -= 1;
    if (flagLeft) startCol -= 1;
    if (flagRight) endCol += 1;
    if (flagBottom) endRow += 1;
  }
}
